pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,664 @@
+	}
+
+	isHome := workspaceType == "Home"
+	if foundLocally, retryAfterSeconds, err := h.searchForReadyWorkspaceInLocalInformers(lcluster.Name, isHome, user.GetName()); foundLocally {
+		if retryAfterSeconds > 0 {
+			// Return a 429 status asking the client to try again after the creationDelay
+			rw.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfterSeconds))
+			http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+			return
+		}
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	} else if err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+
+	// Home or bucket workspace not found in the local informer
+	// Let's try to create it
+
+	// But first check we have the right to do so.
+	attributes := homeWorkspaceAuthorizerAttributes(user, "create")
+
+	if isHome && lcluster.Name != h.getHomeLogicalClusterName(user.GetName()) {
+		// If we're checking a home workspace or home bucket workspace, but not of the consistent user, let's refuse.
+		klog.Errorf("failed to authorize user %q to create a home workspace %q: home workspace can only be created by the user of the home workspace", user.GetName(), lcluster.Name)
+		responsewriters.Forbidden(ctx, attributes, rw, req, authorization.WorkspaceAcccessNotPermittedReason, homeWorkspaceCodecs)
+		return
+	}
+
+	// Test if the user has the right to create his Home workspace when it doesn't exists
+	// => test the create verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+	if decision, reason, err := h.authz.Authorize(
+		request.WithCluster(ctx, request.Cluster{Name: tenancyv1alpha1.RootCluster}),
+		attributes); err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	} else if decision != authorizer.DecisionAllow {
+		responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+		return
+	}
+
+	if retryAfterSeconds, err := h.tryToCreate(ctx, user.GetName(), lcluster.Name, workspaceType); err != nil {
+		responsewriters.ErrorNegotiated(err, errorCodecs, schema.GroupVersion{}, rw, req)
+		return
+	} else {
+		// Return a 429 status asking the client to try again after the creationDelay
+		rw.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfterSeconds))
+		http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+		return
+	}
+}
+
+// reHomeWorkspaceNameDisallowedtChars is the regexp that defines what characters
+// are disallowed in a home workspace name.
+// Home workspace name is derived from the user name, with disallowed characters
+// replaced.
+var reHomeWorkspaceNameDisallowedtChars = regexp.MustCompile("[^a-z0-9-]")
+
+// getHomeLogicalClusterName returns the logicalcluster name of the home workspace for a given user
+// The home workspace logical cluster ancestors are home bucket workspaces whose name is based
+// on the user name sha1 hash.
+func (h *homeWorkspaceHandler) getHomeLogicalClusterName(userName string) logicalcluster.Name {
+	// bucketLevels <= 5
```

Comment:
What's the meaning of configuring these values? When would an admin choose a different value for either the levels or size? What implication would it have? Does an admin have the right information to choose this? Can it change over time?

Reply:
please move these restriction to the constructor doc if it's not already the case, and validate there.

Reply:
done

Reply:
@davidfestal @sttts if this cannot change over time we need to be EXTREMELY clear and careful so admins choose a correct value here.